File: chatuikit/src/main/java/com/cometchat/chatuikit/shared/views/inlineaudiorecorder/InlineAudioRecorderState.java
package com.cometchat.chatuikit.shared.views.inlineaudiorecorder;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

/**
 * Represents the internal state of the inline audio recorder.
 * <p>
 * This class holds all state information for the audio recorder including
 * current status, recording duration, playback position, file path, error messages,
 * and amplitude data for waveform visualization.
 * </p>
 *
 * @see InlineAudioRecorderStatus
 */
public class InlineAudioRecorderState {

    private InlineAudioRecorderStatus status;
    private long duration;
    private long currentPosition;
    private String filePath;
    private String errorMessage;
    private List<Float> amplitudes;

    /**
     * Creates a new InlineAudioRecorderState with default values.
     * Initial state is IDLE with zero duration and position.
     */
    public InlineAudioRecorderState() {
        this.status = InlineAudioRecorderStatus.IDLE;
        this.duration = 0;
        this.currentPosition = 0;
        this.filePath = null;
        this.errorMessage = null;
        this.amplitudes = new ArrayList<>();
    }

    /**
     * Creates a new InlineAudioRecorderState with specified values.
     *
     * @param status          the current recorder status
     * @param duration        the total recorded duration in milliseconds
     * @param currentPosition the current playback position in milliseconds
     * @param filePath        the path to the recorded audio file
     * @param errorMessage    the error message if status is ERROR
     * @param amplitudes      the list of amplitude values for waveform visualization
     */
    public InlineAudioRecorderState(InlineAudioRecorderStatus status, long duration,
                                    long currentPosition, String filePath,
                                    String errorMessage, List<Float> amplitudes) {
        this.status = status;
        this.duration = duration;
        this.currentPosition = currentPosition;
        this.filePath = filePath;
        this.errorMessage = errorMessage;
        this.amplitudes = amplitudes != null ? new ArrayList<>(amplitudes) : new ArrayList<>();
    }

    // Getters and Setters

    public InlineAudioRecorderStatus getStatus() {
        return status;
    }

    public void setStatus(InlineAudioRecorderStatus status) {
        this.status = status;
    }

    public long getDuration() {
        return duration;
    }

    public void setDuration(long duration) {
        this.duration = duration;
    }

    public long getCurrentPosition() {
        return currentPosition;
    }

    public void setCurrentPosition(long currentPosition) {
        this.currentPosition = currentPosition;
    }

    public String getFilePath() {
        return filePath;
    }

    public void setFilePath(String filePath) {
        this.filePath = filePath;
    }

    public String getErrorMessage() {
        return errorMessage;
    }

    public void setErrorMessage(String errorMessage) {
        this.errorMessage = errorMessage;
    }

    public List<Float> getAmplitudes() {
        return amplitudes;
    }

    public void setAmplitudes(List<Float> amplitudes) {
        this.amplitudes = amplitudes != null ? new ArrayList<>(amplitudes) : new ArrayList<>();
    }

    // Helper methods

    /**
     * Checks if the recorder is currently recording.
     *
     * @return true if status is RECORDING, false otherwise
     */
    public boolean isRecording() {
        return status == InlineAudioRecorderStatus.RECORDING;
    }

    /**
     * Checks if the recording is paused.
     *
     * @return true if status is PAUSED, false otherwise
     */
    public boolean isPaused() {
        return status == InlineAudioRecorderStatus.PAUSED;
    }

    /**
     * Checks if the recording is completed.
     *
     * @return true if status is COMPLETED, false otherwise
     */
    public boolean isCompleted() {
        return status == InlineAudioRecorderStatus.COMPLETED;
    }

    /**
     * Checks if the recorder is currently playing back audio.
     *
     * @return true if status is PLAYING, false otherwise
     */
    public boolean isPlaying() {
        return status == InlineAudioRecorderStatus.PLAYING;
    }

    /**
     * Checks if a recording exists (file path is set).
     *
     * @return true if filePath is not null and not empty, false otherwise
     */
    public boolean hasRecording() {
        return filePath != null && !filePath.isEmpty();
    }

    /**
     * Resets all state to initial values.
     * Sets status to IDLE, clears duration, position, file path, error message,
     * and amplitudes list.
     */
    public void reset() {
        this.status = InlineAudioRecorderStatus.IDLE;
        this.duration = 0;
        this.currentPosition = 0;
        this.filePath = null;
        this.errorMessage = null;
        this.amplitudes = new ArrayList<>();
    }

    /**
     * Creates a copy of this state object.
     * The copy is a new instance with the same field values.
     *
     * @return a new InlineAudioRecorderState with identical values
     */
    public InlineAudioRecorderState copy() {
        return new InlineAudioRecorderState(
                this.status,
                this.duration,
                this.currentPosition,
                this.filePath,
                this.errorMessage,
                this.amplitudes
        );
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        InlineAudioRecorderState that = (InlineAudioRecorderState) o;
        return duration == that.duration &&
                currentPosition == that.currentPosition &&
                status == that.status &&
                Objects.equals(filePath, that.filePath) &&
                Objects.equals(errorMessage, that.errorMessage) &&
                Objects.equals(amplitudes, that.amplitudes);
    }

    @Override
    public int hashCode() {
        return Objects.hash(status, duration, currentPosition, filePath, errorMessage, amplitudes);
    }

    @Override
    public String toString() {
        return "InlineAudioRecorderState{" +
                "status=" + status +
                ", duration=" + duration +
                ", currentPosition=" + currentPosition +
                ", filePath='" + filePath + '\'' +
                ", errorMessage='" + errorMessage + '\'' +
                ", amplitudesCount=" + (amplitudes != null ? amplitudes.size() : 0) +
                '}';
    }
}

File: chatuikit-core/src/main/java/com/cometchat/uikit/core/viewmodel/InlineAudioRecorderState.kt
package com.cometchat.uikit.core.viewmodel

/**
 * Immutable state for the Inline Audio Recorder.
 * Contains all state needed for UI rendering.
 *
 * This data class represents the complete state of the inline audio recorder,
 * including recording status, duration, playback position, file path, error message,
 * and amplitude history for waveform visualization.
 *
 * @property status Current recording status (IDLE, RECORDING, PAUSED, COMPLETED, PLAYING, ERROR)
 * @property duration Total recorded duration in milliseconds
 * @property currentPosition Current playback position in milliseconds (used during PLAYING state)
 * @property filePath Path to the recorded audio file (null when no recording exists)
 * @property errorMessage Error message if status is ERROR (null otherwise)
 * @property amplitudes List of stored amplitude values (0.0 to 1.0) captured during recording,
 *                      used for playback waveform visualization
 *
 * @see InlineAudioRecorderStatus
 * @see CometChatInlineAudioRecorderViewModel
 */
data class InlineAudioRecorderState(
    /**
     * Current status of the inline audio recorder.
     * Determines which UI elements are visible and their behavior.
     */
    val status: InlineAudioRecorderStatus = InlineAudioRecorderStatus.IDLE,

    /**
     * Total recorded duration in milliseconds.
     * Updated during recording and preserved after recording completes.
     */
    val duration: Long = 0L,

    /**
     * Current playback position in milliseconds.
     * Only relevant during PLAYING state, updated via polling at 100ms intervals.
     */
    val currentPosition: Long = 0L,

    /**
     * Path to the recorded audio file.
     * Set when recording completes, cleared when recording is deleted.
     */
    val filePath: String? = null,

    /**
     * Error message when status is ERROR.
     * Null when no error has occurred.
     */
    val errorMessage: String? = null,

    /**
     * List of amplitude values captured during recording.
     * Each value is normalized to range [0.0, 1.0].
     * Used for waveform visualization during playback.
     * New amplitudes are appended to the end (bars appear on right, scroll left).
     */
    val amplitudes: List<Float> = emptyList()
) {
    /**
     * Calculates the playback progress as a value between 0.0 and 1.0.
     * Returns 0.0 if duration is 0 to avoid division by zero.
     */
    val playbackProgress: Float
        get() = if (duration > 0) (currentPosition.toFloat() / duration.toFloat()).coerceIn(0f, 1f) else 0f

    /**
     * Returns true if the recorder is in a state where recording can be started.
     */
    val canStartRecording: Boolean
        get() = status == InlineAudioRecorderStatus.IDLE

    /**
     * Returns true if the recorder is in a state where recording can be paused.
     */
    val canPauseRecording: Boolean
        get() = status == InlineAudioRecorderStatus.RECORDING

    /**
     * Returns true if the recorder is in a state where recording can be resumed.
     */
    val canResumeRecording: Boolean
        get() = status == InlineAudioRecorderStatus.PAUSED

    /**
     * Returns true if the recorder is in a state where recording can be stopped.
     */
    val canStopRecording: Boolean
        get() = status == InlineAudioRecorderStatus.RECORDING || status == InlineAudioRecorderStatus.PAUSED

    /**
     * Returns true if the recorder is in a state where playback can be started.
     */
    val canStartPlayback: Boolean
        get() = status == InlineAudioRecorderStatus.COMPLETED

    /**
     * Returns true if the recorder is in a state where playback can be paused.
     */
    val canPausePlayback: Boolean
        get() = status == InlineAudioRecorderStatus.PLAYING

    /**
     * Returns true if the recorder is in a state where the recording can be deleted.
     */
    val canDelete: Boolean
        get() = status != InlineAudioRecorderStatus.IDLE

    /**
     * Returns true if the recorder is in a state where the recording can be submitted.
     */
    val canSubmit: Boolean
        get() = status == InlineAudioRecorderStatus.COMPLETED || 
                status == InlineAudioRecorderStatus.PLAYING ||
                status == InlineAudioRecorderStatus.PAUSED

    /**
     * Returns true if the recorder is actively recording (not paused).
     */
    val isRecording: Boolean
        get() = status == InlineAudioRecorderStatus.RECORDING

    /**
     * Returns true if the recorder is playing back audio.
     */
    val isPlaying: Boolean
        get() = status == InlineAudioRecorderStatus.PLAYING

    /**
     * Returns true if the recorder has an error.
     */
    val hasError: Boolean
        get() = status == InlineAudioRecorderStatus.ERROR
}
